# Pittsburgh Pennsylvania Regional Dataset


## Business Problem


#### Business Problem
 
 In this project the business problem is how we can move to a new neighboorhood in Pittsburgh PA with similer venues providing information about the area along with similer house prices that clients can afford.
  
 This project utalizes machine learning technieqes to provide insights on how homeowners can choose a desired neighberhood based on their needs. This project retrives all schooling, resturant, bar, gym, supermarket and many more entertainment venues to divied these neighboorhoods to colored clusters in maps to provide visual similarities in the city of Pittsburgh, PA.


#### Dataset : Allegheny County Property Sale Transactions

This dataset obtainted from Allegheny County (Pennsylvania) and the City of Pittsburgh both publish their data through the Western Pennsylvania Regional Data Center. The Western Pennsylvania Regional Data Center supports key community initiatives by making public information easier to find and use. The Data Center is managed by the University of Pittsburgh’s Center for Social and Urban Research, and is a partnership of the University, Allegheny County and the City of Pittsburgh.contains data on all Real Property parcels that have sold since 2013 in Allegheny County, PA.

# Methodology

<a href="linkedin.com/pulse/chapter-1-introduction-crisp-dm-framework-data-science-anshul-roy"><img src = "https://media-exp1.licdn.com/dms/image/C4E12AQFtnsyAeMMBMw/article-cover_image-shrink_720_1280/0?e=1591228800&v=beta&t=gQZG09k7GVH_Xc3ODLHO9ndT7WDUvQGAx4PpZ7kox-s" width = 400> </a>

 #  1. Collect Inspection Data

###### This is the first step in this process to Install and import all required libraries and uplade the dateset to start the analysis:

In [1]:
import pandas as pd # library for data analsysis
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)
!conda install -c conda-forge geopy --yes # uncomment this line if you haven't completed the Foursquare API lab
from geopy.geocoders import Nominatim # convert an address into latitude and longitude values
!conda install -c conda-forge folium=0.5.0 --yes # uncomment this line if you haven't completed the Foursquare API lab
import folium # map rendering library
import numpy as np
import datetime as DT
import hmac
from geopy.geocoders import Nominatim
from geopy.distance import vincenty
# import k-means from clustering stage
from sklearn.cluster import KMeans
import os
import datetime as dt # Datetime
import json # library to handle JSON files
from geopy.geocoders import Nominatim # convert an address into latitude and longitude values
import requests # library to handle requests
from pandas.io.json import json_normalize # tranform JSON file into a pandas dataframe
# Matplotlib and associated plotting modules
import matplotlib.cm as cm
import matplotlib.colors as colors

usage: conda-script.py [-h] [-V] command ...
conda-script.py: error: unrecognized arguments: # uncomment this line if you haven't completed the Foursquare API lab
usage: conda-script.py [-h] [-V] command ...
conda-script.py: error: unrecognized arguments: # uncomment this line if you haven't completed the Foursquare API lab


In [2]:
df=pd.read_csv('A.csv')

# 2. Data Analysis

###### Anaysing dataset to better understand the variables

In [3]:
df.dropna(inplace=True)

In [4]:
df['SALEDATE'] = df['SALEDATE'].apply(pd.to_datetime)
df.sort_values(by=['SALEDATE'],ascending=[False],inplace=True)
df.drop(df[df.SALEDATE.dt.year < 2019].index, inplace=True)

In [5]:
df_price = df.groupby(['Address'])['PRICE'].mean().reset_index()
df_price.columns = ['Street', 'Price']
df_affordable = df_price.query("(Price >= 1500000) & (Price <= 2500000)")
df_affordable.dtypes.dropna(inplace=True) 

# 3. Data preparation and preprocessing

#### In this section data mining technique that involves transforming raw data into an understandable format are utalized to prepare data for model creation.

In [6]:
for index, item in df_affordable.iterrows():
    print(f"index: {index}")
    print(f"item: {item}")
    print(f"item.Street only: {item.Street}")

index: 74
item: Street    0 ALPHA DR ALLISON PARK PA 15101
Price                             1.95e+06
Name: 74, dtype: object
item.Street only: 0 ALPHA DR ALLISON PARK PA 15101
index: 79
item: Street    0 AMBER ST PITTSBURGH PA 15206
Price                            1.5e+06
Name: 79, dtype: object
item.Street only: 0 AMBER ST PITTSBURGH PA 15206
index: 215
item: Street    0 BRANDYWINE CT MC DONALD PA 15057
Price                                2.2e+06
Name: 215, dtype: object
item.Street only: 0 BRANDYWINE CT MC DONALD PA 15057
index: 262
item: Street    0 BUTLER PLANK RD GLENSHAW PA 15116
Price                                 2.4e+06
Name: 262, dtype: object
item.Street only: 0 BUTLER PLANK RD GLENSHAW PA 15116
index: 811
item: Street    0 LAKE MARSHALL DR GIBSONIA PA 15044
Price                                 1.85e+06
Name: 811, dtype: object
item.Street only: 0 LAKE MARSHALL DR GIBSONIA PA 15044
index: 857
item: Street    0 LINCOLN HWY NORTH VERSAILLES PA 15137
Price                

In [7]:
from geopy.geocoders import Nominatim
geolocator = Nominatim(timeout=10)
df_affordable['city_coord'] = df_affordable['Street'].apply(geolocator.geocode)

C:\Users\Maryam\Anaconda3\lib\site-packages\ipykernel_launcher.py:2: DeprecationWarning: Using Nominatim with the default "geopy/1.21.0" `user_agent` is strongly discouraged, as it violates Nominatim's ToS https://operations.osmfoundation.org/policies/nominatim/ and may possibly cause 403 and 429 HTTP errors. Please specify a custom `user_agent` with `Nominatim(user_agent="my-application")` or by overriding the default `user_agent`: `geopy.geocoders.options.default_user_agent = "my-application"`. In geopy 2.0 this will become an exception.
  
C:\Users\Maryam\Anaconda3\lib\site-packages\ipykernel_launcher.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  This is separate from the ipykernel package so we can avoid doing imports until


In [8]:
df_affordable.dropna(inplace=True)

C:\Users\Maryam\Anaconda3\lib\site-packages\ipykernel_launcher.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  """Entry point for launching an IPython kernel.


In [9]:
df_affordable['city_coord'] = df_affordable['city_coord'].apply(lambda x: (x.latitude, x.longitude))

C:\Users\Maryam\Anaconda3\lib\site-packages\ipykernel_launcher.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  """Entry point for launching an IPython kernel.


In [10]:
df_affordable[['Latitude', 'Longitude']] = df_affordable['city_coord'].apply(pd.Series)

C:\Users\Maryam\Anaconda3\lib\site-packages\pandas\core\frame.py:3494: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  self[k1] = value[k2]


In [11]:
df = df_affordable.drop(columns=['city_coord'])

In [12]:
address = 'PITTSBURGH, PA'

geolocator = Nominatim(timeout=10)
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The geograpical coordinate of PITTSBURGH are {}, {}.'.format(latitude, longitude))

C:\Users\Maryam\Anaconda3\lib\site-packages\ipykernel_launcher.py:3: DeprecationWarning: Using Nominatim with the default "geopy/1.21.0" `user_agent` is strongly discouraged, as it violates Nominatim's ToS https://operations.osmfoundation.org/policies/nominatim/ and may possibly cause 403 and 429 HTTP errors. Please specify a custom `user_agent` with `Nominatim(user_agent="my-application")` or by overriding the default `user_agent`: `geopy.geocoders.options.default_user_agent = "my-application"`. In geopy 2.0 this will become an exception.
  This is separate from the ipykernel package so we can avoid doing imports until


The geograpical coordinate of PITTSBURGH are 40.4416941, -79.9900861.


In [13]:
map_Pittsburgh = folium.Map(location=[latitude, longitude], zoom_start=11)

# add markers to map
for lat, lng, price, street in zip(df['Latitude'], df['Longitude'], df['Price'], df['Street']):
    label = '{}, {}'.format(street, price)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_Pittsburgh)  
    
map_Pittsburgh

In [14]:
CLIENT_ID = '3ULFNWEKKLJWG00R5RSRRP1GZDBDCQNIOERJLQUV5QYIMI0L' # Foursquare ID
CLIENT_SECRET = 'APQ3I5EEZZWLD2KKQJ4H35ZLVEJXUJGJANNTPIDH305QXYFV' # Foursquare Secret
VERSION = '20181206' # Foursquare API version

print('Your credentails:')
print('CLIENT_ID: ' + CLIENT_ID)
print('CLIENT_SECRET:' + CLIENT_SECRET)

Your credentails:
CLIENT_ID: 3ULFNWEKKLJWG00R5RSRRP1GZDBDCQNIOERJLQUV5QYIMI0L
CLIENT_SECRET:APQ3I5EEZZWLD2KKQJ4H35ZLVEJXUJGJANNTPIDH305QXYFV


# 4. Modeling

##### This process is produced a descriptive map and models and their relationships between various types of information that are to be stored in a database.

In [15]:
def getNearbyVenues(names, latitudes, longitudes, radius=500, LIMIT=100):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Street', 
                  'Street Latitude', 
                  'Street Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)

In [16]:
location_venues = getNearbyVenues(names=df['Street'],
                                   latitudes=df['Latitude'],
                                   longitudes=df['Longitude']
                                  )

0 ALPHA DR ALLISON PARK PA 15101
0 AMBER ST PITTSBURGH PA 15206
0 BRANDYWINE CT MC DONALD PA 15057
0 BUTLER PLANK RD GLENSHAW PA 15116
0 LAKE MARSHALL DR GIBSONIA PA 15044
0 LINCOLN HWY NORTH VERSAILLES PA 15137
0 MARYLAND DR GLENSHAW PA 15116
0 PENN AVE PITTSBURGH PA 15206
0 SHADY AVE PITTSBURGH PA 15232
0 STEUBENVILLE PIKE PITTSBURGH PA 15205
0 SUTTER RD GLENSHAW PA 15116
0 WOODLAND RD SEWICKLEY PA 15143
1005 BEAVER GRADE RD CORAOPOLIS PA 15108
104 AUDUBON RD SEWICKLEY PA 15143
1074 DEVON RD PITTSBURGH PA 15213
1101 SHADY AVE PITTSBURGH PA 15232
118 52ND ST PITTSBURGH PA 15201
121 9TH ST PITTSBURGH PA 15222
125 51ST ST PITTSBURGH PA 15201
1273 WASHINGTON PIKE BRIDGEVILLE PA 15017
1281 WASHINGTON PIKE BRIDGEVILLE PA 15017
1297 WASHINGTON PIKE BRIDGEVILLE PA 15017
12980 PERRY HWY WEXFORD PA 15090
1343 WEST LIBERTY AVE PITTSBURGH PA 15226
1374 FREEPORT RD PITTSBURGH PA 15238
158 NORTH DR PITTSBURGH PA 15238
1649 PENN AVE PITTSBURGH PA 15222
1651 SUTTER RD GLENSHAW PA 15116
1825 CENTRE A

In [18]:
from matplotlib.ticker import FuncFormatter
import matplotlib.pyplot as plt
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt

In [20]:
location_venues.groupby('Street').count()

Street Latitude  \
Street                                                          
0 ALPHA DR ALLISON PARK PA 15101                            2   
0 AMBER ST PITTSBURGH PA 15206                             46   
0 BRANDYWINE CT MC DONALD PA 15057                          2   
0 BUTLER PLANK RD GLENSHAW PA 15116                         7   
0 LINCOLN HWY NORTH VERSAILLES PA 15137                     6   
0 MARYLAND DR GLENSHAW PA 15116                             4   
0 PENN AVE PITTSBURGH PA 15206                             52   
0 SHADY AVE PITTSBURGH PA 15232                            57   
0 STEUBENVILLE PIKE PITTSBURGH PA 15205                     3   
0 SUTTER RD GLENSHAW PA 15116                               2   
0 WOODLAND RD SEWICKLEY PA 15143                            2   
1005 BEAVER GRADE RD CORAOPOLIS PA 15108                    2   
1074 DEVON RD PITTSBURGH PA 15213                          10   
1101 SHADY AVE PITTSBURGH PA 15232                         10   
118 52ND ST PITTSBURGH PA 15201                            30   
121 9TH ST PITTSBURGH PA 15222                            100   
125 51ST ST PITTSBURGH PA 15201                            19   
1273 WASHINGTON PIKE BRIDGEVILLE PA 15017                  36   
1281 WASHINGTON PIKE BRIDGEVILLE PA 15017                  36   
1297 WASHINGTON PIKE BRIDGEVILLE PA 15017                  36   
12980 PERRY HWY WEXFORD PA 15090                           24   
1343 WEST LIBERTY AVE PITTSBURGH PA 15226                   6   
1374 FREEPORT RD PITTSBURGH PA 15238                       42   
158 NORTH DR PITTSBURGH PA 15238                           29   
1649 PENN AVE PITTSBURGH PA 15222                          48   
1651 SUTTER RD GLENSHAW PA 15116                            3   
1825 CENTRE AVE PITTSBURGH PA 15219                        12   
1901 CENTRE AVE PITTSBURGH PA 15219                         3   
1918 TUSTIN ST PITTSBURGH PA 15219                          2   
200 AMBER ST PITTSBURGH PA 15206                           27   
200 ELLIOTT RD MONROEVILLE PA 15146                         4   
200 JAMES PL MONROEVILLE PA 15146                           2   
201 FREEPORT RD PITTSBURGH PA 15215                        13   
2055 LEBANON CHURCH RD WEST MIFFLIN PA 15122                4   
209 9TH ST PITTSBURGH PA 15222                            100   
211 AMBER ST PITTSBURGH PA 15206                           32   
2116 BROWNSVILLE RD PITTSBURGH PA 15210                    18   
2400 MARKET PLACE BLVD CORAOPOLIS PA 15108                  2   
2400 SMALLMAN ST PITTSBURGH PA 15222                       53   
2401 SMALLMAN ST PITTSBURGH PA 15222                       53   
2908 SMALLMAN ST PITTSBURGH PA 15201                       16   
304 JUMONVILLE ST PITTSBURGH PA 15219                       2   
318 GRANT ST SEWICKLEY PA 15143                            17   
326 HAYS AVE PITTSBURGH PA 15210                           16   
3301 BEECHWOOD BLVD PITTSBURGH PA 15217                     7   
331 PENN AVE PITTSBURGH PA 15221                           14   
4025 ALPHA DR ALLISON PARK PA 15101                         2   
4035 ALPHA DR ALLISON PARK PA 15101                         2   
4065 ALPHA DR ALLISON PARK PA 15101                         2   
4075 ALPHA DR ALLISON PARK PA 15101                         2   
4085 ALPHA DR ALLISON PARK PA 15101                         2   
422 ELFINWILD RD ALLISON PARK PA 15101                      4   
427 WOODLAND RD SEWICKLEY PA 15143                          2   
4338 BIGELOW BLVD PITTSBURGH PA 15213                      22   
446 BEAVER RD SEWICKLEY PA 15143                            3   
460 SOUTH AVE VERONA PA 15147                               8   
4976 MCKNIGHT RD PITTSBURGH PA 15237                        3   
5036 AMBERSON PL PITTSBURGH PA 15232                        7   
516 BACKBONE RD SEWICKLEY PA 15143                          1   
5526 WALNUT ST PITTSBURGH PA 15232                         40   
5565 NORTHUMBERL

In [21]:
print('There are {} uniques categories.'.format(len(location_venues['Venue Category'].unique())))

There are 209 uniques categories.


In [22]:
location_venues.shape

(1484, 7)

In [23]:
# one hot encoding
venues_onehot = pd.get_dummies(location_venues[['Venue Category']], prefix="", prefix_sep="")
venues_onehot['Street'] = location_venues['Street'] 
fixed_columns = [venues_onehot.columns[-1]] + list(venues_onehot.columns[:-1])
venues_onehot = venues_onehot[fixed_columns]
venues_onehot.head()

Street  Accessories Store  Advertising Agency  \
0  0 ALPHA DR ALLISON PARK PA 15101                  0                   0   
1  0 ALPHA DR ALLISON PARK PA 15101                  0                   0   
2    0 AMBER ST PITTSBURGH PA 15206                  0                   0   
3    0 AMBER ST PITTSBURGH PA 15206                  0                   0   
4    0 AMBER ST PITTSBURGH PA 15206                  0                   0   

   Airport  American Restaurant  Antique Shop  Arcade  Argentinian Restaurant  \
0        0                    0             0       0                       0   
1        0                    0             0       0                       0   
2        0                    0             0       0                       0   
3        0                    0             0       0                       0   
4        0                    0             0       0                       0   

   Art Gallery  Arts & Crafts Store  Asian Restaurant  Athletics & Sports  \
0            0                    0                 0                   0   
1            0                    0                 0                   0   
2            0                    0                 0                   0   
3            0                    0                 1                   0   
4            1                    0                 0                   0   

   Auto Garage  Auto Workshop  BBQ Joint  Bagel Shop  Bakery  Bank  Bar  \
0            0              0          0           0       0     0    0   
1            0              0          0           0       0     0    0   
2            0              0          0           0       0     0    0   
3            0              0          0           0       0     0    0   
4            0              0          0           0       0     0    0   

   Baseball Field  Beer Bar  Beer Garden  Beer Store  Big Box Store  Bistro  \
0               0         0            0           0              0       0   
1               0         0            0           0              0       0   
2               0         0            0           0              0       0   
3               0         0            0           0              0       0   
4               0         0            0           0              0       0   

   Boat or Ferry  Bookstore  Boutique  Bowling Alley  Breakfast Spot  Brewery  \
0              0          0         0              0               0        0   
1              0          0         0              0               0        0   
2              0          0         0              0               0        0   
3              0          0         0              0               0        0   
4              0          0         0              0               0        0   

   Bridge  Bubble Tea Shop  Building  Burger Joint  Bus Line  Bus Station  \
0       0                0         0             0         0            0   
1       0                0         0             0         0            0   
2       0                0         0             0         0            0   
3       0                0         0             0         0            0   
4       0                0         0             0         0            0   

   Bus Stop  Business Service  Butcher  Café  Candy Store  \
0         0                 0        0     0            0   
1         0                 0        0     0            0   
2         0                 0        0     0            0   
3         0                 0        0     0            0   
4         0                 0        0     0            0   

   Caribbean Restaurant  Chinese Restaurant  Climbing Gym  Clothing Store  \
0                     0                   0             0               0   
1                     0                   0             0               0   
2                     0                   0             0               0   
3                     0                   0             0  

In [24]:
Pitt_grouped = venues_onehot.groupby('Street').mean().reset_index()

In [25]:
Pitt_grouped.shape

(78, 210)

In [26]:
# What are the top 5 venues/facilities nearby profitable real estate investments?#

num_top_venues = 5

for hood in Pitt_grouped['Street']:
    print("----"+hood+"----")
    temp = Pitt_grouped[Pitt_grouped['Street'] == hood].T.reset_index()
    temp.columns = ['venue','freq']
    temp = temp.iloc[1:]
    temp['freq'] = temp['freq'].astype(float)
    temp = temp.round({'freq': 2})
    print(temp.sort_values('freq', ascending=False).reset_index(drop=True).head(num_top_venues))
    print('\n')

----0 ALPHA DR ALLISON PARK PA 15101----
               venue  freq
0               Park   0.5
1        Pizza Place   0.5
2  Accessories Store   0.0
3             Museum   0.0
4        Music Store   0.0


----0 AMBER ST PITTSBURGH PA 15206----
                  venue  freq
0           Coffee Shop  0.11
1         Grocery Store  0.07
2                 Hotel  0.04
3  Gym / Fitness Center  0.04
4             Pet Store  0.04


----0 BRANDYWINE CT MC DONALD PA 15057----
               venue  freq
0            Brewery   0.5
1         Steakhouse   0.5
2  Accessories Store   0.0
3      Movie Theater   0.0
4        Music Store   0.0


----0 BUTLER PLANK RD GLENSHAW PA 15116----
                  venue  freq
0  Gym / Fitness Center  0.14
1              Pharmacy  0.14
2           Gas Station  0.14
3          Skating Rink  0.14
4                  Bank  0.14


----0 LINCOLN HWY NORTH VERSAILLES PA 15137----
                 venue  freq
0                  Gym  0.17
1          Video Store  0.17
2     

              venue  freq
0     Deli / Bodega  0.12
1       Auto Garage  0.06
2       Bus Station  0.06
3  Storage Facility  0.06
4       Flower Shop  0.06


----304 JUMONVILLE ST PITTSBURGH PA 15219----
               venue  freq
0               Park   0.5
1                Bar   0.5
2  Accessories Store   0.0
3             Museum   0.0
4        Music Store   0.0


----318 GRANT ST SEWICKLEY PA 15143----
                  venue  freq
0           Coffee Shop  0.12
1     Accessories Store  0.06
2  Fast Food Restaurant  0.06
3          Soccer Field  0.06
4         Movie Theater  0.06


----326 HAYS AVE PITTSBURGH PA 15210----
               venue  freq
0     Discount Store  0.12
1  Convenience Store  0.12
2             Bakery  0.06
3        Pizza Place  0.06
4           Pharmacy  0.06


----3301 BEECHWOOD BLVD PITTSBURGH PA 15217----
                  venue  freq
0          Intersection  0.29
1                Museum  0.14
2  Marijuana Dispensary  0.14
3               Airport  0.14
4      

In [27]:
# Define a function to return the most common venues/facilities nearby real estate investments#

def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

In [28]:
num_top_venues = 10
indicators = ['st', 'nd', 'rd']
# create columns according to number of top venues
columns = ['Street']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

In [29]:
# create a new dataframe
venues_sorted = pd.DataFrame(columns=columns)
venues_sorted['Street'] = Pitt_grouped['Street']

for ind in np.arange(Pitt_grouped.shape[0]):
    venues_sorted.iloc[ind, 1:] = return_most_common_venues(Pitt_grouped.iloc[ind, :], num_top_venues)



In [30]:
venues_sorted.head()

Street 1st Most Common Venue  \
0         0 ALPHA DR ALLISON PARK PA 15101           Pizza Place   
1           0 AMBER ST PITTSBURGH PA 15206           Coffee Shop   
2       0 BRANDYWINE CT MC DONALD PA 15057            Steakhouse   
3      0 BUTLER PLANK RD GLENSHAW PA 15116           Supermarket   
4  0 LINCOLN HWY NORTH VERSAILLES PA 15137   American Restaurant   

  2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0                  Park                   Zoo     Electronics Store   
1         Grocery Store             Pet Store  Gym / Fitness Center   
2               Brewery     Food & Drink Shop                  Food   
3              Pharmacy  Gym / Fitness Center    Salon / Barbershop   
4        Sandwich Place              Pharmacy           Video Store   

  5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
0           Flower Shop           Fish Market     Fish & Chips Shop   
1   American Restaurant    Salon / Barbershop                 Hotel   
2           Flower Shop           Fish Market     Fish & Chips Shop   
3          Skating Rink           Gas Station                  Bank   
4                   Gym             Pawn Shop                   Zoo   

        8th Most Common Venue 9th Most Common Venue  \
0  Financial or Legal Service  Fast Food Restaurant   
1                    Pharmacy          Liquor Store   
2  Financial or Legal Service  Fast Food Restaurant   
3                         Zoo     Electronics Store   
4                     Exhibit  Ethiopian Restaurant   

       10th Most Common Venue  
0              Farmers Market  
1        Ethiopian Restaurant  
2              Farmers Market  
3  Financial or Legal Service  
4                 Event Space

In [32]:
venues_sorted.shape

(78, 11)

In [33]:
Pitt_grouped.shape

(78, 210)

In [34]:
Pitt_grouped=df

In [35]:
#Distribute in 5 Clusters

# set number of clusters
kclusters = 5

Pitt_grouped_clustering = Pitt_grouped.drop('Street', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(Pitt_grouped_clustering)

# check cluster labels generated for each row in the dataframe
kmeans.labels_[0:50]

array([0, 1, 3, 2, 4, 0, 2, 2, 1, 3, 2, 1, 2, 1, 0, 1, 0, 4, 0, 2, 2, 2,
       4, 0, 1, 1, 3, 2, 1, 0, 3, 4, 1, 2, 1, 1, 1, 4, 1, 2, 4, 1, 3, 4,
       3, 1, 4, 1, 4, 0])

In [36]:
#Dataframe to include Clusters

Pitt_grouped_clustering=df
Pitt_grouped_clustering.head()

Street      Price   Latitude  Longitude
74       0 ALPHA DR ALLISON PARK PA 15101  1950000.0  40.551668 -79.987459
79         0 AMBER ST PITTSBURGH PA 15206  1500000.0  40.460333 -79.931427
215    0 BRANDYWINE CT MC DONALD PA 15057  2200000.0  40.708323 -75.226708
262   0 BUTLER PLANK RD GLENSHAW PA 15116  2400000.0  40.538425 -79.962543
811  0 LAKE MARSHALL DR GIBSONIA PA 15044  1850000.0  40.601984 -79.994453

In [37]:
Pitt_grouped_clustering.shape

(83, 4)

In [38]:
df.shape

(83, 4)

In [40]:
df.dtypes

Street        object
Price        float64
Latitude     float64
Longitude    float64
dtype: object

In [41]:
# add clustering labels
Pitt_grouped_clustering['Cluster Labels'] = kmeans.labels_

# merge london_grouped with london_data to add latitude/longitude for each neighborhood
Pitt_grouped_clustering = Pitt_grouped_clustering.join(venues_sorted.set_index('Street'), on='Street')

Pitt_grouped_clustering.head(30) # check the last columns!

Street      Price   Latitude  \
74              0 ALPHA DR ALLISON PARK PA 15101  1950000.0  40.551668   
79                0 AMBER ST PITTSBURGH PA 15206  1500000.0  40.460333   
215           0 BRANDYWINE CT MC DONALD PA 15057  2200000.0  40.708323   
262          0 BUTLER PLANK RD GLENSHAW PA 15116  2400000.0  40.538425   
811         0 LAKE MARSHALL DR GIBSONIA PA 15044  1850000.0  40.601984   
857      0 LINCOLN HWY NORTH VERSAILLES PA 15137  2077032.0  40.380975   
949              0 MARYLAND DR GLENSHAW PA 15116  2400000.0  40.533236   
1193              0 PENN AVE PITTSBURGH PA 15206  2400000.0  40.461037   
1364             0 SHADY AVE PITTSBURGH PA 15232  1597471.0  40.459410   
1436     0 STEUBENVILLE PIKE PITTSBURGH PA 15205  2150000.0  40.440420   
1455               0 SUTTER RD GLENSHAW PA 15116  2400000.0  40.543959   
1612            0 WOODLAND RD SEWICKLEY PA 15143  1500000.0  40.553231   
1961    1005 BEAVER GRADE RD CORAOPOLIS PA 15108  2462500.0  40.492324   
2824           104 AUDUBON RD SEWICKLEY PA 15143  1550000.0  40.575082   
3421           1074 DEVON RD PITTSBURGH PA 15213  2000000.0  40.445601   
3866          1101 SHADY AVE PITTSBURGH PA 15232  1597471.0  40.449776   
5148             118 52ND ST PITTSBURGH PA 15201  2030000.0  40.479302   
5620              121 9TH ST PITTSBURGH PA 15222  1800000.0  40.444585   
6241             125 51ST ST PITTSBURGH PA 15201  1900000.0  40.480068   
6526   1273 WASHINGTON PIKE BRIDGEVILLE PA 15017  2425000.0  40.363924   
6591   1281 WASHINGTON PIKE BRIDGEVILLE PA 15017  2425000.0  40.363924   
6685   1297 WASHINGTON PIKE BRIDGEVILLE PA 15017  2425000.0  40.363924   
6688            12980 PERRY HWY WEXFORD PA 15090  1800000.0  40.609766   
7468   1343 WEST LIBERTY AVE PITTSBURGH PA 15226  2060000.0  40.410930   
7714        1374 FREEPORT RD PITTSBURGH PA 15238  1550000.0  40.487204   
9647            158 NORTH DR PITTSBURGH PA 15238  1525000.0  40.441694   
10255          1649 PENN AVE PITTSBURGH PA 15222  2125000.0  40.449320   
10280           1651 SUTTER RD GLENSHAW PA 15116  2400000.0  40.544346   
11438        1825 CENTRE AVE PITTSBURGH PA 15219  1600000.0  40.442672   
11747        1901 CENTRE AVE PITTSBURGH PA 15219  2040000.0  40.443241   

       Longitude  Cluster Labels       1st Most Common Venue  \
74    -79.987459               0                 Pizza Place   
79    -79.931427               1                 Coffee Shop   
215   -75.226708               3                  Steakhouse   
262   -79.962543               2                 Supermarket   
811   -79.994453               4                         NaN   
857   -79.814566               0         American Restaurant   
949   -79.965067               2               Garden Center   
1193  -79.923766               2                 Coffee Shop   
1364  -79.921497               1                 Coffee Shop   
1436  -80.081290               3               Women's Store   
1455  -79.973079               2                     Brewery   
1612  -80.189283               1                     Brewery   
1961  -80.195542               2        Gym / Fitness Center   
2824  -80.137584               1                         NaN   
3421  -79.940564               0                 Coffee Shop   
3866  -79.920326               1                Intersection   
5148  -79.952858               0                         Bar   
5620  -79.998619               4                       Hotel   
6241  -79.957353               0                         Bar   
6526  -80.109809               2                 Pizza Place   
6591  -80.109809               2                 Pizza Place   
6685  -80.109809               2                 Pizza Place   
6688  -80.053192               4                Liquor Store   
7468  -80.013477               0          Light Rail Station   
7714  -79.885359               1              Cosmetics Shop   
9647  -79.990086               1                       Hotel   
10255 -79.986868   

In [44]:
# Create Map

map_clusters = folium.Map(location=[latitude, longitude], zoom_start=11)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i+x+(i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(Pitt_grouped_clustering['Latitude'], Pitt_grouped_clustering['Longitude'], Pitt_grouped_clustering['Street'], Pitt_grouped_clustering['Cluster Labels']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[cluster-1],
        fill=True,
        fill_color=rainbow[cluster-1],
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters

In [45]:
Pitt_grouped_clustering.loc[Pitt_grouped_clustering['Cluster Labels'] == 0, Pitt_grouped_clustering.columns[[1] + list(range(5, Pitt_grouped_clustering.shape[1]))]].head()


Price 1st Most Common Venue 2nd Most Common Venue  \
74    1950000.0           Pizza Place                  Park   
857   2077032.0   American Restaurant        Sandwich Place   
3421  2000000.0           Coffee Shop     Indian Restaurant   
5148  2030000.0                   Bar               Brewery   
6241  1900000.0                   Bar                 River   

     3rd Most Common Venue   4th Most Common Venue  5th Most Common Venue  \
74                     Zoo       Electronics Store            Flower Shop   
857               Pharmacy             Video Store                    Gym   
3421           Bus Station              Food Truck  Performing Arts Venue   
5148      Tapas Restaurant  Thrift / Vintage Store     Chinese Restaurant   
6241      Asian Restaurant                 Brewery             Smoke Shop   

     6th Most Common Venue      7th Most Common Venue  \
74             Fish Market          Fish & Chips Shop   
857              Pawn Shop                        Zoo   
3421     College Cafeteria  College Academic Building   
5148    Salon / Barbershop              Garden Center   
6241               Gay Bar                     Bakery   

           8th Most Common Venue   9th Most Common Venue  \
74    Financial or Legal Service    Fast Food Restaurant   
857                      Exhibit    Ethiopian Restaurant   
3421            Sculpture Garden                   Hotel   
5148                 Record Shop               Juice Bar   
6241                 Flower Shop  Thrift / Vintage Store   

     10th Most Common Venue  
74           Farmers Market  
857             Event Space  
3421    Arts & Crafts Store  
5148            Flower Shop  
6241          Grocery Store

In [46]:
Pitt_grouped_clustering.loc[Pitt_grouped_clustering['Cluster Labels'] == 1, Pitt_grouped_clustering.columns[[1] + list(range(5, Pitt_grouped_clustering.shape[1]))]].head()


Price 1st Most Common Venue 2nd Most Common Venue  \
79    1500000.0           Coffee Shop         Grocery Store   
1364  1597471.0           Coffee Shop    Mexican Restaurant   
1612  1500000.0               Brewery           Art Gallery   
2824  1550000.0                   NaN                   NaN   
3866  1597471.0          Intersection           Bus Station   

     3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
79               Pet Store  Gym / Fitness Center   American Restaurant   
1364   American Restaurant           Pizza Place                   Bar   
1612                   Zoo  Ethiopian Restaurant                  Food   
2824                   NaN                   NaN                   NaN   
3866                  Park          Home Service           Art Gallery   

     6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
79      Salon / Barbershop                 Hotel              Pharmacy   
1364          Noodle House     French Restaurant        Ice Cream Shop   
1612           Flower Shop           Fish Market     Fish & Chips Shop   
2824                   NaN                   NaN                   NaN   
3866           Coffee Shop                   Zoo  Ethiopian Restaurant   

           9th Most Common Venue 10th Most Common Venue  
79                  Liquor Store   Ethiopian Restaurant  
1364          Chinese Restaurant           Cocktail Bar  
1612  Financial or Legal Service   Fast Food Restaurant  
2824                         NaN                    NaN  
3866                 Fish Market      Fish & Chips Shop

In [47]:
Pitt_grouped_clustering.loc[Pitt_grouped_clustering['Cluster Labels'] == 2, Pitt_grouped_clustering.columns[[1] + list(range(5, Pitt_grouped_clustering.shape[1]))]].head()


Price 1st Most Common Venue 2nd Most Common Venue  \
262   2400000.0           Supermarket              Pharmacy   
949   2400000.0         Garden Center            Beer Store   
1193  2400000.0           Coffee Shop    Mexican Restaurant   
1455  2400000.0               Brewery       Thai Restaurant   
1961  2462500.0  Gym / Fitness Center                   Gym   

     3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
262   Gym / Fitness Center    Salon / Barbershop          Skating Rink   
949                   Bank          Liquor Store                   Zoo   
1193           Pizza Place   American Restaurant  Gym / Fitness Center   
1455                   Zoo           Event Space                  Food   
1961                   Zoo     Electronics Store           Flower Shop   

     6th Most Common Venue 7th Most Common Venue       8th Most Common Venue  \
262            Gas Station                  Bank                         Zoo   
949             Food Court                  Food                 Flower Shop   
1193         Grocery Store                   Bar              Ice Cream Shop   
1455           Flower Shop           Fish Market           Fish & Chips Shop   
1961           Fish Market     Fish & Chips Shop  Financial or Legal Service   

           9th Most Common Venue      10th Most Common Venue  
262            Electronics Store  Financial or Legal Service  
949                  Fish Market           Fish & Chips Shop  
1193                Noodle House           French Restaurant  
1455  Financial or Legal Service        Fast Food Restaurant  
1961        Fast Food Restaurant              Farmers Market

In [48]:
Pitt_grouped_clustering.loc[Pitt_grouped_clustering['Cluster Labels'] == 3, Pitt_grouped_clustering.columns[[1] + list(range(5, Pitt_grouped_clustering.shape[1]))]].head()


Price 1st Most Common Venue 2nd Most Common Venue  \
215    2200000.0            Steakhouse               Brewery   
1436   2150000.0         Women's Store   Rental Car Location   
10255  2125000.0                   Bar                Bakery   
11888  2200000.0           Gas Station                   Bar   
15664  2200000.0                   Bar           Coffee Shop   

        3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
215         Food & Drink Shop                  Food           Flower Shop   
1436   Furniture / Home Store                   Zoo     Electronics Store   
10255             Coffee Shop           Art Gallery          Gourmet Shop   
11888                     Zoo           Event Space                  Food   
15664           Grocery Store                Bakery    Seafood Restaurant   

      6th Most Common Venue 7th Most Common Venue       8th Most Common Venue  \
215             Fish Market     Fish & Chips Shop  Financial or Legal Service   
1436            Fish Market     Fish & Chips Shop  Financial or Legal Service   
10255           Pizza Place         Grocery Store                     Gay Bar   
11888           Flower Shop           Fish Market           Fish & Chips Shop   
15664                 Diner           Pizza Place                Cocktail Bar   

            9th Most Common Venue 10th Most Common Venue  
215          Fast Food Restaurant         Farmers Market  
1436         Fast Food Restaurant         Farmers Market  
10255                       Diner         Sandwich Place  
11888  Financial or Legal Service   Fast Food Restaurant  
15664                 Art Gallery           Gourmet Shop

In [49]:
Pitt_grouped_clustering.loc[Pitt_grouped_clustering['Cluster Labels'] == 4, Pitt_grouped_clustering.columns[[1] + list(range(5, Pitt_grouped_clustering.shape[1]))]].head()


Price 1st Most Common Venue 2nd Most Common Venue  \
811    1850000.0                   NaN                   NaN   
5620   1800000.0                 Hotel   American Restaurant   
6688   1800000.0          Liquor Store  Fast Food Restaurant   
12081  1850000.0                   NaN                   NaN   
13210  1750000.0                 Hotel               Theater   

      3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
811                     NaN                   NaN                   NaN   
5620                Theater           Coffee Shop                   Bar   
6688      Food & Drink Shop   Rental Car Location            Shoe Store   
12081                   NaN                   NaN                   NaN   
13210           Coffee Shop    Italian Restaurant                   Bar   

      6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
811                     NaN                   NaN                   NaN   
5620             Restaurant         Deli / Bodega    Mexican Restaurant   
6688              Pet Store      Sushi Restaurant    Salon / Barbershop   
12081                   NaN                   NaN                   NaN   
13210   American Restaurant            Restaurant                 Plaza   

      9th Most Common Venue 10th Most Common Venue  
811                     NaN                    NaN  
5620     Italian Restaurant            Pizza Place  
6688     Chinese Restaurant           Optical Shop  
12081                   NaN                    NaN  
13210         Deli / Bodega                    Gym

# Analytical Results

### Business Solution:

Pittsburgh has become a world-class city, home prices are still lower than the US national average. The median home value in Pittsburgh is $164,317. The median list price per square foot in Pittsburgh is $144, which is higher than the Pittsburgh Metro average of $125. The median price of homes currently listed in Pittsburgh is $224,900 while the median price of homes that sold is $187,400. The median rent price in Pittsburgh is $1,450, which is higher than the Pittsburgh Metro median of $1,225. Machine learning tools provide essential insights to homebuyers and businesses to assist them with decision making based on their needs. This project solves the business problem with clustering 5 neighborhoods, recommending all sorts of venues surrounding these spots in Pittsburgh, PA. We recommended profitable venues according to amenities and essential facilities surrounding such venues i.e. elementary schools, high schools, hospitals & grocery stores. 

In this process first we gathered data from Allegheny County (Pennsylvania) and the City of Pittsburgh which gathered their data through the Western Pennsylvania Regional Data Center. after data collection, we analyzed the data to gather statistical information on Pittsburgh housing data. Then we used geolocator library to import longitude and latitude of the properties and used Foursquare API interface to collect information on venues based on the location of the property. By shaping this comprehensive dataset, we are able to provide profitable insight on the location desired houses in Pittsburgh’s area. Unsupervised machine learning, k-mean clustering method utilized to cluster 5 neighborhoods that have similar characteristics in the city of Pittsburgh. 

In this project we conclude that clusters that have natural amenities are in high demand than the ones that are surrounded with Plazas including restaurants and shopping centers. Cluster 2 and 4 are evident for this fact. In this study we also see that cluster 0 in East Pittsburgh is the best area for profitable purchases as it shows similar characteristics in cluster 2 and 4. One might target underpriced investments in these areas of Pittsburgh in order to make a business affair.